hal/src/rtl872x/exflash_hal.cpp

```diff
@@ -364,31 +370,298 @@ static bool isSecureOtpMode(uint32_t normalContent) {
 
 #if MODULE_FUNCTION != MOD_FUNC_BOOTLOADER
 
+// GD25 flash specific definitions
+
+#define FIFO_SIZE                           64
+
+#define bfoSPI_FLASH_AUTO_LEN_DUM           ((uint32_t)     0)
+#define bfwSPI_FLASH_AUTO_LEN_DUM           ((uint32_t)    16)
+#define bfoSPI_FLASH_SR_TXE                 ((uint32_t)    5)
+#define bfoSPI_FLASH_SR_BUSY                ((uint32_t)    0)
+
+#define DEF_RD_TUNING_DUMMY_CYCLE           0x2
+
+#define DW_BIT_MASK(__bfws) ((uint32_t) ((__bfws) == 32) ?                  \
+    0x0 : (0x1 << (__bfws)))
+
+#define DW_BIT_MASK_WIDTH(__bfws, __bits) ((uint32_t) ((__bfws) == 32) ?    \
+    0xFFFFFFFF : (((1 << (__bits)) - 1) << (__bfws)))
+
+#define DW_BITS_SET_VAL(__datum, __bfws, __val, bit_num)                    \
+    ((__datum) = ((uint32_t) (__datum) &                                    \
+                ~DW_BIT_MASK_WIDTH(__bfws, bit_num)) |                      \
+    ((__val << (__bfws)) & DW_BIT_MASK_WIDTH(__bfws, bit_num)))
+
+#define CMD_ADDR_FORMAT(cmd, addr) (((cmd) & 0x000000ff) |                  \
+                ((addr & 0x000000ff) << 24) |                               \
+                ((addr & 0x0000ff00) << 8) |                                \
+                ((addr & 0x00ff0000) >> 8))
+
+#define DW_BIT_GET_UNSHIFTED(__datum, __bfws)                               \
+    ((uint32_t) ((__datum) & DW_BIT_MASK(__bfws)))
+
+enum spi_flash_byte_num {
+    DATA_BYTE = 0,
+    DATA_HALF = 1,
+    DATA_WORD = 2
+};
+
+__attribute__((section(".ram.text"), noinline))
```

Comment:
Any difference if we run out of RAM or PSRAM?

Reply:
Nope. These new functions should always be called after boot up. Would you like to remove the restriction to save more RAM sapce?

Reply:
Just to avoid unnecessary attributes for readability. If they work fine out of regular PSRAM-copied `.text`, I'd just leave them there.